### PR TITLE
HIP CPU

### PR DESCRIPTION
**Pull request by dc-dc-dc**
Got it working but it is brutally slow right now, with 16 cores it takes almost an hour just to run `test/test_ops.py`

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/runtime/ops_hip.py               63              20
------------------------------------------------------------
lines added in the tinygrad folder: 43
```

**Comment by prusnak** [comment]
`'utf-8'` is the default encoding for `encode()` and `decode()`, so you can just call `.encode()`/`.decode()`

**Comment by dc-dc-dc** [comment]
current bottleneck is that it has to compile the runtime for each kernel as HIP-CPU is missing `hipModuleLaunchKernel` and instead uses an inline macro to launch the kernel.

**Comment by dc-dc-dc** [comment]
Found something that is closer to whats needed `hip::detail::launch_kernel_from_so` but currently getting seg faults when trying to use it

**Comment by aseok** [comment]
@dc-dc-dc Is there any of already proposed bounties relates to your pr?

**Comment by geohot** [comment]
Locked the bounty for you.

File: tinygrad/runtime/ops_hip.py [hunk 1]
```diff
@@ -16,53 +22,90 @@
...
       asm = early_exec((["/opt/rocm/llvm/bin/llvm-objdump", '-d', '-'], prg))
       print('\n'.join([x for x in asm.decode('utf-8').split("\n") if 's_code_end' not in x]))
-
-    module = hip.hipModuleLoadData(prg)
-    self.prg = hip.hipModuleGetFunction(module, name)
+    if HIP_CPU:
+      import tempfile, hashlib, os, subprocess
+      fn = f"{tempfile.gettempdir()}/clang_{hashlib.md5(prg.encode('utf-8')).hexdigest()}.so"
+      if not os.path.exists(fn):
+        subprocess.check_output(args=('clang -g -std=c++20 --stdlib=libstdc++ -shared -O2 -Wall -Werror -x c++ -ltbb -ltbbmalloc -fPIC --rtlib=compiler-rt - -o '+fn+'.tmp').split(), input=prg.encode('utf-8'))
```
Comment (prusnak) [inline tinygrad/runtime/ops_hip.py:49]: this is a race condition - add process id to tmp file name - see `ops_clang.py` to see how it is done:

https://github.com/tinygrad/tinygrad/blob/a293c18d345a45ff7a315c8fb5da17437c129949/tinygrad/runtime/ops_clang.py#L38

**Review by prusnak — CHANGES_REQUESTED** [review]

**Comment by wozeparrot** [comment]
Update on this?

**Comment by dc-dc-dc** [comment]
Will close and let someone else continue, I still think `hip::detail::launch_kernel_from_so` is the right path forward as using `hipLaunchKernelGGL` requires rebuilding the runtime which takes way too long and times out in CI

Tracking #1700

**Comment by nimlgen** [comment]
`hip::detail::launch_kernel_from_so` does not support any kernel arguments at all. Call to `launch` [passes empty args tuple](https://github.com/ROCm-Developer-Tools/HIP-CPU/blob/master/src/include/hip/detail/api.hpp#L636) and seems there is a [comment](https://github.com/ROCm-Developer-Tools/HIP-CPU/blob/master/src/include/hip/detail/api.hpp#L638).

Compilation time of "HIP-CPU" include-only templated library is huge. I am not sure we can do anything with this, since even `blockIdx`, which is inside each kernel, can take up to 5-20s to compile.

**Pull request closed without merging**